lib/server/sse_handler.dart

```diff
@@ -28,16 +28,37 @@ class SseConnection extends StreamChannelMixin<String> {
   /// Outgoing messages to the Browser client.
   final _outgoingController = StreamController<String>();
 
-  final Sink _sink;
+  Sink _sink;
+
+  /// How long to wait after a connection drops before considering it closed.
+  final Duration _keepAlive;
+
+  /// Whether the connection is in the timeout period waiting for a reconnect.
+  bool _isTimingOut = false;
+
+  /// The subscription that passes messages outgoing messages to the sink. This
+  /// will be paused during the timeout/reconnect period.
+  StreamSubscription _outgoingStreamSubscription;
 
   final _closedCompleter = Completer<void>();
 
-  SseConnection(this._sink) {
-    _outgoingController.stream.listen((data) {
+  SseConnection(this._sink, {Duration keepAlive}) : _keepAlive = keepAlive {
+    _outgoingStreamSubscription = _outgoingController.stream.listen((data) {
       if (!_closedCompleter.isCompleted) {
-        // JSON encode the message to escape new lines.
-        _sink.add('data: ${json.encode(data)}\n');
-        _sink.add('\n');
+        try {
+          // JSON encode the message to escape new lines.
+          _sink.add('data: ${json.encode(data)}\n');
+          _sink.add('\n');
+        } catch (StateError) {
+          if (_keepAlive == null) {
+            rethrow;
+          }
+          // If we got here then the sink may have closed but the stream.onDone
+          // hasn't fired yet, so pause the subscription, re-queue the message
+          // and handle the error as a disconnect.
+          _handleDisconnect();
+          _outgoingController.add(data);
```

Comment:
An alternative would be to use a `StreamQueue` and use `peak` / `next` to conditionally move forward. That should guarantee order.

Reply:
That sounds good - the order does concern me. However I'm not sure how this would work - without the stream, how would we fire this code?

I see there's a `sync` argument in [StreamController](https://api.dartlang.org/stable/2.7.0/dart-async/StreamController/StreamController.html) that could fix this (this code would fire on the first event, so there wouldn't be others to get out of order), though I don't know if it'd introduce any other side-effects?

Reply:
I'm thinking something like this:

```Dart
var outgoingStreamQueue = StreamQueue(_outgoingController.stream);
    while (await outgoingStreamQueue.hasNext) {
      var peek = await outgoingStreamQueue.peek;
      try {
        send(peek);
        await outgoingStreamQueue.next;
      } catch (e) {
        // Handle error and don't call `next`
      }
    }
```

Reply:
> `await outgoingStreamQueue.hasNext`

Aha, I missed that - this looks sensible to me. I'll try and make it work Monday (as well as test on GitPod to confirm it actually solves the issue). Thanks!

Reply:
I think this code would burn CPU when an exception occurs (if we don't call `next`, `hasNext` will return `true` and just sit in a tight loop). The existing code pauses the stream so it wouldn't enter the handler again.

A fix could be to `await Future.delayed(const Duration(milliseconds: x))` though it feels a bit weird (and if the delay is too large, we introduce latency, and too low and we burn more CPU).

Can you think of a better way?

Reply:
I went with a wait period of 200ms for now and added a test for in-order messages. I think this all works, though I don't know how you feel about the 200ms retry (I thought about using exponential backoff, but we could easily end up with a large delay that would blow the keepAlive period and a simple check 5 times per second doesn't seem too wasteful for something we'd expect to occur within a number of seconds).

LMK what you think!

```dart
      // If we're in a KeepAlive timeout, there's nowhere to send messages so
      // wait a short period and check again.
      if (isInKeepAlivePeriod) {
        await Future.delayed(const Duration(milliseconds: 200));
        continue;
      }
```